### PR TITLE
Part 5 of Async Compaction: Ensure Cleaner and Archiver do not delete file-slices and workload marked for compaction

### PR DESCRIPTION
**Pull request by bvaradar**
Ensure Cleaner and Archiver do not delete file-slices and workload marked for compaction

This addresses safety concerns in https://github.com/bvaradar/hudi/pull/6

### CODE REVIEW

**Comment by bvaradar** [comment]
@n3nash : Can you take a pass at this

File: hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java [hunk 1]
```diff
@@ -679,6 +679,42 @@ public void deleteSavepoint(String savepointTime) {
...
+   * Delete a compaction request that is pending.
+   *
+   * NOTE - This is an Admin operation.
+   * With async compaction, this is expected to be called with async compaction and ingestion shutdown.
+   * Otherwise, async compactor could fail with errors
+   *
+   * @param compactionTime - delete the compaction time
+   * @return
+   */
+  public void deletePendingCompaction(String compactionTime) {
```
Comment (n3nash) [inline hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java]: Is this a tool ? We should move this to hoodie-cli, if it is needed here, it should be a private method.

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java [hunk 2]
```diff
@@ -48,6 +55,7 @@
 
   private final TableFileSystemView fileSystemView;
   private final HoodieTimeline commitTimeline;
+  private final Map<String, HoodieCompactionOperation> fileIdToPendingCompactionOperations;
```
Comment (n3nash) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java]: Can we avoid defining this as a member variable ? Seems out of place...

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java [hunk 3]
```diff
@@ -16,6 +16,9 @@
```
Comment (n3nash) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java:16]: In general, the logic of pending compactions file slices removal is very interleaved with the cleaner code, can you separate them out ? Something like getAllEligibleFileSlices() where one function removed all the file slices from the list that should not be compacted and the other just does what this code used to do ? There is an extra loop over the file slices but I feel that will be ok given clean is a one time operation for some batches anyways. Thoughts ?

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java [hunk 4]
```diff
@@ -173,6 +189,49 @@ private boolean deleteArchivedInstants(List<HoodieInstant> archivedInstants) {
         throw new HoodieIOException("Failed to delete archived instant " + archivedInstant, e);
       }
     }
+
+    // Remove older meta-data from auxiliary path too
+    Optional<HoodieInstant> youngestCommitted =
```
Comment (n3nash) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: Rename of oldestCommitted or latestCommitted based on what you mean by youngest..

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java [hunk 5]
```diff
@@ -173,6 +189,49 @@ private boolean deleteArchivedInstants(List<HoodieInstant> archivedInstants) {
...
+  }
+
+  /**
+   * Remove older instants from auxiliary meta folder
+   *
+   * @param instant Hoodie Instant
+   * @return success if all eligible file deleted successfully
+   * @throws IOException in case of error
+   */
+  private boolean deleteAllInstantsOlderorEqualsInAuxMetaFolder(HoodieInstant instant)
```
Comment (n3nash) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: Have we added tests for this ?

**Comment by n3nash** [comment]
Haven't been able to go over the test classes for this or any PR, FYI

File: hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java (hunk 1 above)
Comment (vinothchandar) [reply hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java]: +1

File: hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java [hunk 6]
```diff
@@ -679,6 +679,42 @@ public void deleteSavepoint(String savepointTime) {
...
+    } else {
+      logger.error("No compaction present " + compactionTime);
+      throw new IllegalArgumentException("No compaction present " + compactionTime);
+    }
+    logger.info("Compaction " + compactionTime + " deleted");
+  }
+
   /**
    * Rollback the state to the savepoint. WARNING: This rollsback recent commits and deleted data
    * files. Queries accessing the files will mostly fail. This should be done during a downtime.
```
Comment (vinothchandar) [inline hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:722]: but then, similar stuff (admin, tooling) is already in the  HoodieWriteClient..

Comment (vinothchandar) [reply hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:722]: Can we move these into a subclass of HoodieWriteClient - called HoodieAdminWriteClient ?  Just throwing it out there

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java [hunk 7]
```diff
@@ -224,6 +238,34 @@ private String getLatestVersionBeforeCommit(List<FileSlice> fileSliceList,
       earliestCommitToRetain = commitTimeline
           .nthInstant(commitTimeline.countInstants() - commitsRetained);
     }
-    return earliestCommitToRetain;
+
+    // ensure oldest commit time needed for pending compaction is retained
+    final Optional<HoodieInstant> oldestCommitNeededForCompaction =
```
Comment (vinothchandar) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java]: oldestInstantTimeNeededForCompaction ?

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java [hunk 8]
```diff
@@ -143,23 +148,34 @@ public boolean archiveIfRequired() {
     //TODO (na) : Add a way to return actions associated with a timeline and then merge/unify
     // with logic above to avoid Stream.concats
     HoodieTimeline commitTimeline = table.getCompletedCommitTimeline();
+    Optional<HoodieInstant> oldestPendingCompactionInstant =
```
Comment (vinothchandar) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java:153]: can we use "earliest" "latest" to stay consistent with rest of the code?

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java (hunk 4 above)
Comment (vinothchandar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: +1  left a similar comment above

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java [hunk 9]
```diff
@@ -173,6 +189,49 @@ private boolean deleteArchivedInstants(List<HoodieInstant> archivedInstants) {
         throw new HoodieIOException("Failed to delete archived instant " + archivedInstant, e);
       }
     }
+
+    // Remove older meta-data from auxiliary path too
+    Optional<HoodieInstant> youngestCommitted =
+        archivedInstants.stream()
+            .filter(i -> {
+              return i.isCompleted()
+                  && (i.getAction().equals(COMMIT_ACTION) || (i.getAction().equals(DELTA_COMMIT_ACTION)));
```
Comment (vinothchandar) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: the inclusion of delta_commit_action also is to facilitate move of delta commit/rollback metadata?

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java (hunk 5 above)
Comment (vinothchandar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: +1

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java [hunk 10]
```diff
@@ -173,6 +189,49 @@ private boolean deleteArchivedInstants(List<HoodieInstant> archivedInstants) {
...
+  private boolean deleteAllInstantsOlderorEqualsInAuxMetaFolder(HoodieInstant instant)
+      throws IOException {
+    List<HoodieInstant> instants =
+        HoodieTableMetaClient.scanHoodieInstantsFromFileSystem(metaClient.getFs(),
+            new Path(metaClient.getMetaAuxiliaryPath()),
+            HoodieActiveTimeline.INCLUDED_EXTENSIONS_IN_ACTIVE_TIMELINE);
+
+    List<HoodieInstant> instantsToBeDeleted =
+        instants.stream().filter(instant1 -> {
+          return instant1.getTimestamp().compareTo(instant.getTimestamp()) <= 0;
```
Comment (vinothchandar) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: is nt there a method to compare instants to each other more nicely?

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java [hunk 11]
```diff
@@ -173,6 +189,49 @@ private boolean deleteArchivedInstants(List<HoodieInstant> archivedInstants) {
...
+        }).collect(Collectors.toList());
+
+    boolean success = true;
+    for (HoodieInstant instant1 : instantsToBeDeleted) {
```
Comment (vinothchandar) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: rename: deleteInstant

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java (hunk 5 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: New Test cases and validations are added in TestHoodieCommitArchiveLog

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java (hunk 3 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java:16]: Lets revisit further refactoring holistically after we make a push on finishing up pending features related to async compaction and usability.  I dont want to introduce new bugs late in the game

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java (hunk 2 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java]: We can revisit refactoring later.

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java (hunk 4 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: Done

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java (hunk 7 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java]: this code is removed.

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java (hunk 9 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: Yes

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java (hunk 10 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: Fixed.

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java (hunk 11 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java]: Fixed

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java (hunk 8 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java:153]: Fixed

File: hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java (hunk 1 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java]: I have made it a private method. Its used for the rollback case

File: hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java (hunk 6 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:722]: This is a bit involved as the method is used by rollback logic which in turn is referenced by write/compaction case. HoodieWriteClient needs to be cleaned up anyways and we can tackle it as part of that effort.  Opened an issue for this : https://github.com/uber/hudi/issues/432

Comment (n3nash) [reply hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:722]: sounds good to me

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java (hunk 3 above)
Comment (n3nash) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java:16]: I'm okay with that.

**Comment by bvaradar** [comment]
Committed as part of Consolidated diff

**Pull request closed without merging**